File: mcp_server/memory/dedup.py
"""Semantic deduplication for memories."""

import logging
import os
from dataclasses import dataclass
from typing import Any

import numpy as np

from .embedder import Embedder
from .store import ValkeyStore

logger = logging.getLogger(__name__)


@dataclass
class DuplicateMatch:
    """A memory that is semantically near-identical to a candidate."""

    key: str
    content: str
    similarity: float
    project: str | None
    state: str
    created_at: str


def check_duplicate(
    store: ValkeyStore,
    namespace: str,
    vector: np.ndarray,
    content: str,
    threshold: float | None = None,
    project_filter: str | None = None,
) -> DuplicateMatch | None:
    """Search namespace for a near-identical memory.

    Args:
        store: ValkeyStore instance.
        namespace: Namespace to search within.
        vector: Pre-computed embedding vector (reused from remember()).
        content: The new content text (for comparison logging).
        threshold: Similarity threshold (default from DEDUP_SIMILARITY_THRESHOLD env).
        project_filter: Optional project scope.

    Returns:
        The best matching duplicate above threshold, or None.
    """
    if threshold is None:
        threshold = float(os.getenv("DEDUP_SIMILARITY_THRESHOLD", "0.92"))

    results = store.search(namespace, vector, top_k=5)
    if not results:
        return None

    for doc in results:
        # Skip archived/deleted
        state = doc.get("state", "active")
        if state in ("archived", "deleted"):
            continue

        # Project filter
        if project_filter:
            doc_project = doc.get("project") or doc.get("project_name")
            if doc_project != project_filter:
                continue

        # Convert cosine distance to similarity
        similarity = 1.0 - float(doc.get("similarity_score", "1.0"))
        if similarity >= threshold:
            return DuplicateMatch(
                key=doc.get("key", ""),
                content=doc.get("content", ""),
                similarity=similarity,
                project=doc.get("project") or doc.get("project_name"),
                state=state,
                created_at=doc.get("created_at", ""),
            )

    return None


def find_all_duplicates(
    store: ValkeyStore,
    embedder: Embedder,
    namespace: str = "episodic",
    threshold: float | None = None,
    project_filter: str | None = None,
    max_keys: int = 2000,
) -> list[list[dict[str, Any]]]:
    """Scan all memories in a namespace and return clusters of duplicates.

    Re-embeds all content and performs pairwise comparison.
    Capped at max_keys to prevent excessive runtime.

    Args:
        store: ValkeyStore instance.
        embedder: Embedder instance.
        namespace: Namespace to scan.
        threshold: Similarity threshold (default from env).
        project_filter: Optional project filter.
        max_keys: Maximum keys to scan.

    Returns:
        List of duplicate clusters. Each cluster is a list of memory dicts.
    """
    if threshold is None:
        threshold = float(os.getenv("DEDUP_SIMILARITY_THRESHOLD", "0.92"))

    prefix = f"mem:{namespace}:"
    keys = store.scan_prefix(prefix)
    if not keys:
        return []

    if len(keys) > max_keys:
        logger.warning(
            "Dedup scan capped at %d keys (total: %d)", max_keys, len(keys)
        )
        keys = keys[:max_keys]

    all_data = store.get_multi(keys)

    # Filter and collect content for embedding
    valid_entries: list[tuple[str, dict[str, Any]]] = []
    for key, data in zip(keys, all_data):
        if data is None:
            continue
        if data.get("state") in ("archived", "deleted"):
            continue
        if project_filter:
            doc_project = data.get("project") or data.get("project_name")
            if doc_project != project_filter:
                continue
        valid_entries.append((key, data))

    if len(valid_entries) < 2:
        return []

    # Batch embed all content
    texts = [data.get("content", "") for _, data in valid_entries]
    vectors = embedder.embed_batch(texts)
    vectors_array = np.array(vectors)

    # Pairwise cosine similarity (vectors are normalised, so dot product = cosine)
    similarity_matrix = vectors_array @ vectors_array.T

    # Union-find for clustering
    n = len(valid_entries)
    parent = list(range(n))

    def find(x: int) -> int:
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    def union(a: int, b: int) -> None:
        ra, rb = find(a), find(b)
        if ra != rb:
            parent[ra] = rb

    for i in range(n):
        for j in range(i + 1, n):
            if similarity_matrix[i, j] >= threshold:
                union(i, j)

    # Group into clusters
    clusters_map: dict[int, list[int]] = {}
    for i in range(n):
        root = find(i)
        clusters_map.setdefault(root, []).append(i)

    # Only return clusters with 2+ members
    clusters: list[list[dict[str, Any]]] = []
    for indices in clusters_map.values():
        if len(indices) < 2:
            continue
        cluster = []
        for idx in indices:
            key, data = valid_entries[idx]
            cluster.append({
                "key": key,
                "content": data.get("content", "")[:200],
                "project": data.get("project") or data.get("project_name"),
                "state": data.get("state", "active"),
                "created_at": data.get("created_at"),
            })
        clusters.append(cluster)

    # Sort clusters by size descending
    clusters.sort(key=len, reverse=True)
    return clusters

File: crates/omnimem-engine/src/dedup.rs
//! Semantic duplicate detection (`memory/dedup.py`).

use std::collections::HashMap;

use omnimem_core::Namespace;
use omnimem_store::SearchFilter;
use serde_json::{Map, Value, json};
use tracing::warn;

use crate::pyfmt::take_chars;
use crate::{Engine, Result};

const MAX_DEDUP_KEYS: usize = 2000;

pub struct DuplicateMatch {
    pub key: String,
    pub content: String,
    pub similarity: f64,
}

fn doc_project(fields: &omnimem_store::Fields) -> Option<&str> {
    fields
        .get("project")
        .filter(|p| !p.is_empty())
        .or_else(|| fields.get("project_name").filter(|p| !p.is_empty()))
        .map(String::as_str)
}

impl Engine {
    /// The first near-identical live memory in the namespace, if any.
    pub fn check_duplicate(
        &self,
        namespace: Namespace,
        vector: &[f32],
        project_filter: Option<&str>,
    ) -> Result<Option<DuplicateMatch>> {
        let hits = self
            .store
            .search(namespace, vector, 5, &SearchFilter::default(), None)?;
        for hit in hits {
            let state = hit
                .fields
                .get("state")
                .map(String::as_str)
                .unwrap_or("active");
            if matches!(state, "archived" | "deleted") {
                continue;
            }
            if let Some(project) = project_filter.filter(|p| !p.is_empty())
                && doc_project(&hit.fields) != Some(project)
            {
                continue;
            }
            let similarity = 1.0 - f64::from(hit.distance);
            if similarity >= self.config.dedup_threshold {
                return Ok(Some(DuplicateMatch {
                    key: hit.key,
                    content: hit.fields.get("content").cloned().unwrap_or_default(),
                    similarity,
                }));
            }
        }
        Ok(None)
    }

    /// Clusters of near-identical memories in a namespace, largest first.
    pub(crate) fn find_all_duplicates(
        &self,
        namespace: Namespace,
        threshold: Option<f64>,
        project_filter: Option<&str>,
    ) -> Result<Vec<Value>> {
        let threshold = threshold.unwrap_or(self.config.dedup_threshold);
        let mut keys = self.store.scan_prefix(&format!("mem:{namespace}:"))?;
        if keys.is_empty() {
            return Ok(Vec::new());
        }
        if keys.len() > MAX_DEDUP_KEYS {
            warn!(
                cap = MAX_DEDUP_KEYS,
                total = keys.len(),
                "dedup scan capped"
            );
            keys.truncate(MAX_DEDUP_KEYS);
        }
        let rows = self.store.get_fields_multi(
            &keys,
            &["content", "state", "project", "project_name", "created_at"],
        )?;
        let entries: Vec<(String, omnimem_store::Fields)> = keys
            .into_iter()
            .zip(rows)
            .filter_map(|(key, row)| row.map(|r| (key, r)))
            .filter(|(_, r)| {
                !matches!(
                    r.get("state").map(String::as_str),
                    Some("archived" | "deleted")
                )
            })
            .filter(|(_, r)| match project_filter.filter(|p| !p.is_empty()) {
                Some(project) => doc_project(r) == Some(project),
                None => true,
            })
            .collect();
        if entries.len() < 2 {
            return Ok(Vec::new());
        }

        let entry_keys: Vec<String> = entries.iter().map(|(k, _)| k.clone()).collect();
        let mut vectors = self.store.get_vectors_multi(&entry_keys);
        let missing: Vec<usize> = (0..vectors.len())
            .filter(|i| vectors[*i].is_none())
            .collect();
        if !missing.is_empty() {
            let texts: Vec<&str> = missing
                .iter()
                .map(|i| {
                    entries[*i]
                        .1
                        .get("content")
                        .map(String::as_str)
                        .unwrap_or("")
                })
                .collect();
            for (i, v) in missing.iter().zip(self.embed_many(&texts)?) {
                vectors[*i] = Some(v);
            }
        }
        let vectors: Vec<Vec<f32>> = vectors.into_iter().map(Option::unwrap_or_default).collect();

        let n = entries.len();
        let mut parent: Vec<usize> = (0..n).collect();
        fn find(parent: &mut [usize], mut x: usize) -> usize {
            while parent[x] != x {
                parent[x] = parent[parent[x]];
                x = parent[x];
            }
            x
        }
        let mut pairs: Vec<(usize, usize, f64)> = Vec::new();
        for i in 0..n {
            for j in (i + 1)..n {
                let sim: f32 = vectors[i].iter().zip(&vectors[j]).map(|(a, b)| a * b).sum();
                let sim = f64::from(sim);
                if sim >= threshold {
                    pairs.push((i, j, sim));
                }
            }
        }
        for &(i, j, _) in &pairs {
            let (ri, rj) = (find(&mut parent, i), find(&mut parent, j));
            if ri != rj {
                parent[ri] = rj;
            }
        }
        let mut max_sim: HashMap<usize, f64> = HashMap::new();
        for &(i, _, sim) in &pairs {
            let root = find(&mut parent, i);
            let best = max_sim.entry(root).or_insert(0.0);
            if sim > *best {
                *best = sim;
            }
        }
        let mut order: Vec<usize> = Vec::new();
        let mut members: HashMap<usize, Vec<usize>> = HashMap::new();
        for i in 0..n {
            let root = find(&mut parent, i);
            members
                .entry(root)
                .or_insert_with(|| {
                    order.push(root);
                    Vec::new()
                })
                .push(i);
        }
        let mut clusters: Vec<(usize, Value)> = Vec::new();
        for root in order {
            let indices = &members[&root];
            if indices.len() < 2 {
                continue;
            }
            let memories: Vec<Value> = indices
                .iter()
                .map(|&idx| {
                    let (key, data) = &entries[idx];
                    let mut m = Map::new();
                    m.insert("key".into(), key.as_str().into());
                    m.insert(
                        "content".into(),
                        take_chars(data.get("content").map(String::as_str).unwrap_or(""), 200)
                            .into(),
                    );
                    m.insert(
                        "project".into(),
                        doc_project(data).map_or(Value::Null, Value::from),
                    );
                    m.insert(
                        "state".into(),
                        data.get("state")
                            .map(String::as_str)
                            .unwrap_or("active")
                            .into(),
                    );
                    m.insert(
                        "created_at".into(),
                        data.get("created_at")
                            .map_or(Value::Null, |c| Value::from(c.as_str())),
                    );
                    Value::Object(m)
                })
                .collect();
            let root_now = find(&mut parent, root);
            clusters.push((
                indices.len(),
                json!({
                    "memories": memories,
                    "max_similarity": max_sim.get(&root_now).copied().unwrap_or(0.0),
                }),
            ));
        }
        clusters.sort_by_key(|c| std::cmp::Reverse(c.0));
        Ok(clusters.into_iter().map(|(_, c)| c).collect())
    }
}
